### PR TITLE
[PT FE] [23325] Add aten::masked_select support for pytorch models.

### PR DESCRIPTION
**Pull request by YutingGao7**
### Details:
 - Add aten::masked_select support for pytorch models.

### Tickets:
 - [*ticket-id*](https://github.com/openvinotoolkit/openvino/issues/23325)

### CODE REVIEW

**Comment by YutingGao7** [comment]
Hi openvino team,
I need some help for two issues that I encountered when I ran `python -m pytest pytorch_tests/test_masked_select.py`.

1. tests with GPU failed with:
```
E       RuntimeError: Exception from src/inference/src/cpp/core.cpp:109:
E       Exception from src/inference/src/dev/core_impl.cpp:566:
E       Device with "GPU" name is not registered in the OpenVINO Runtime
```

2. tests with CPU failed with:
```
______________ TestMaskedSelect.test_masked_select[ ie_device:CPU - precision:FP16 - mask_dtype:<class 'bool'> - input_dtype:<class 'numpy.int32'> - mask_select:ones ] ______________

self = <test_masked_select.TestMaskedSelect object at 0x154a3a590>, mask_select = 'ones', mask_dtype = <class 'bool'>, input_dtype = <class 'numpy.int32'>, ie_device = 'CPU'
precision = 'FP16', ir_version = 11

    @pytest.mark.parametrize(
        "mask_select", ['zeros', 'ones', 'random'])
    @pytest.mark.parametrize("input_dtype", [np.float32, np.float64, int, np.int32])
    @pytest.mark.parametrize("mask_dtype", [bool])  # np.float32 incorrectly casted to bool
    @pytest.mark.nightly
    @pytest.mark.precommit
    def test_masked_select(self, mask_select, mask_dtype, input_dtype, ie_device, precision, ir_version):
>       self._test(*self.create_model(),
                   ie_device, precision, ir_version,
                   kwargs_to_prepare_input={'mask_select': mask_select, 'mask_dtype': mask_dtype, "input_dtype": input_dtype})

pytorch_tests/test_masked_select.py:44: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pytorch_tests/pytorch_layer_test_class.py:141: in _test
    compiled = core.compile_model(converted_model, ie_device)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Core: available plugins[CPU]>
model = <Model: 'Model66'
inputs[
<ConstOutput: names[x] shape[?,?] type: i32>,
<ConstOutput: names[mask] shape[?,?] type: boolean>
]
outputs[
<ConstOutput: names[] shape[...] type: i32>
]>
device_name = 'CPU', config = None

    def compile_model(
        self,
        model: Union[Model, str, Path],
        device_name: Optional[str] = None,
        config: Optional[dict] = None,
    ) -> CompiledModel:
        """Creates a compiled model.
    
        Creates a compiled model from a source Model object or
        reads model and creates a compiled model from IR / ONNX / PDPD / TF and TFLite file.
        This can be more efficient than using read_model + compile_model(model_in_memory_object) flow,
        especially for cases when caching is enabled and cached model is available.
        If device_name is not specified, the default OpenVINO device will be selected by AUTO plugin.
        Users can create as many compiled models as they need, and use them simultaneously
        (up to the limitation of the hardware resources).
    
        :param model: Model acquired from read_model function or a path to a model in IR / ONNX / PDPD /
                      TF and TFLite format.
        :type model: Union[openvino.runtime.Model, str, pathlib.Path]
        :param device_name: Optional. Name of the device to load the model to. If not specified,
                            the default OpenVINO device will be selected by AUTO plugin.
        :type device_name: str
        :param config: Optional dict of pairs:
                       (property name, property value) relevant only for this load operation.
        :type config: dict, optional
        :return: A compiled model.
        :rtype: openvino.runtime.CompiledModel
        """
        if device_name is None:
            return CompiledModel(
                super().compile_model(model, {} if config is None else config),
            )
    
        return CompiledModel(
>           super().compile_model(model, device_name, {} if config is None else config),
        )
E       RuntimeError: Exception from src/inference/src/cpp/core.cpp:109:
E       Exception from src/inference/src/dev/plugin.cpp:54:
E       Exception from src/plugins/intel_cpu/src/node.cpp:90:
E       Unexpected: CPU plug-in doesn't support GatherND operation with dynamic rank. Operation name: aten::masked_select/GatherND
```

File: src/plugins/intel_cpu/thirdparty/onednn
Comment (rkazants) [inline src/plugins/intel_cpu/thirdparty/onednn:1]: please revert this change for onednn dependency

Comment (YutingGao7) [reply src/plugins/intel_cpu/thirdparty/onednn:1]: reverted

File: tests/layer_tests/pytorch_tests/test_masked_select.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    @pytest.mark.parametrize(
+        "mask_select", ['zeros', 'ones', 'random'])
+    @pytest.mark.parametrize("input_dtype", [np.float32, np.float64, int, np.int32])
+    @pytest.mark.parametrize("mask_dtype", [bool])  # np.float32 incorrectly casted to bool
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    def test_masked_select(self, mask_select, mask_dtype, input_dtype, ie_device, precision, ir_version):
+        self._test(*self.create_model(),
+                   ie_device, precision, ir_version,
+                   kwargs_to_prepare_input={'mask_select': mask_select, 'mask_dtype': mask_dtype, "input_dtype": input_dtype})
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_masked_select.py:46]:
```suggestion
    @pytest.mark.nightly
    @pytest.mark.precommit
    def test_masked_select(self, mask_select, input_dtype, ie_device, precision, ir_version):
        self._test(*self.create_model(),
                   ie_device, precision, ir_version,
                   kwargs_to_prepare_input={'mask_select': mask_select, 'mask_dtype': bool, "input_dtype": input_dtype})
```

File: tests/layer_tests/pytorch_tests/test_masked_select.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+    @pytest.mark.skipif(parse_version(torch.__version__) >= parse_version("2.1.0"), reason="pytorch 2.1 and above does not support nonboolean mask")
+    @pytest.mark.parametrize(
+        "mask_select", ['zeros', 'ones', 'random'])
+    @pytest.mark.parametrize("input_dtype", [np.float32, np.float64, int, np.int32])
+    @pytest.mark.parametrize("mask_dtype", [np.uint8, np.int32])  # np.float32 incorrectly casted to bool
```
[the remark this answers is not in the document]
Comment (YutingGao7) [reply tests/layer_tests/pytorch_tests/test_masked_select.py:53]: I was just following test case in test_masked_fill.
https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/pytorch_tests/test_masked_fill.py#L67

Should I keep it or remove it?

**Comment by YutingGao7** [comment]
Hi @mvafin,
It looks like the failure is about the outputs' shapes. However, masked_select's output shape depends on the data.

https://pytorch.org/docs/stable/generated/torch.masked_select.html

What prevents cpu to support dynamic shapes? From the code, it looks like only the Result can be dynamic.

```
E       RuntimeError: Exception from src/inference/src/cpp/core.cpp:109:
E       Exception from src/inference/src/dev/plugin.cpp:54:
E       Exception from src/plugins/intel_cpu/src/node.cpp:90:
E       Unexpected: CPU plug-in doesn't support GatherND operation with dynamic rank. Operation name: aten::masked_select/GatherND
```
In plugins/intel_cpu/src/node.cpp:90, it requires shapes on outputs.
```
    if (typeStr != "Result" && typeStr != "Assign") {
        if (op->get_output_size() == 0) {
            OPENVINO_THROW("Node with type '", typeStr, "' and name '", name, "' does not have any outputs.");
        }
        for (size_t i = 0; i < op->get_output_size(); i++) {
            const auto &shape = op->get_output_partial_shape(i);
            if (shape.rank().is_dynamic()) {
                OPENVINO_THROW("Unexpected: CPU plug-in doesn't support ",
                               getTypeStr(),
                               " operation with dynamic rank. Operation name: ",
                               getName());
            }

            bool isScalar = shape.rank().get_length() == 0;
            outputShapes.emplace_back(isScalar ? ov::PartialShape{1} : shape);
            originalOutputPrecisions.emplace_back(op->get_output_element_type(i));
        }

        childEdges.reserve(outputShapes.size());
    }

```

```
self = <Core: available plugins[CPU]>
model = <Model: 'Model69'
inputs[
<ConstOutput: names[x] shape[1,10] type: i32>,
<ConstOutput: names[mask] shape[1,10] type: boolean>
]
outputs[
<ConstOutput: names[] shape[...] type: i32>
]>
device_name = 'CPU', config = None
```

**Comment by mlukasze** [comment]
> Hi @mvafin, It looks like the failure is about the outputs' shapes. However, masked_select's output shape depends on the data.
> 
> https://pytorch.org/docs/stable/generated/torch.masked_select.html
> 
> What prevents cpu to support dynamic shapes? From the code, it looks like only the Result can be dynamic.
> 
> ```
> E       RuntimeError: Exception from src/inference/src/cpp/core.cpp:109:
> E       Exception from src/inference/src/dev/plugin.cpp:54:
> E       Exception from src/plugins/intel_cpu/src/node.cpp:90:
> E       Unexpected: CPU plug-in doesn't support GatherND operation with dynamic rank. Operation name: aten::masked_select/GatherND
> ```
> 
> In plugins/intel_cpu/src/node.cpp:90, it requires shapes on outputs.
> 
> ```
>     if (typeStr != "Result" && typeStr != "Assign") {
>         if (op->get_output_size() == 0) {
>             OPENVINO_THROW("Node with type '", typeStr, "' and name '", name, "' does not have any outputs.");
>         }
>         for (size_t i = 0; i < op->get_output_size(); i++) {
>             const auto &shape = op->get_output_partial_shape(i);
>             if (shape.rank().is_dynamic()) {
>                 OPENVINO_THROW("Unexpected: CPU plug-in doesn't support ",
>                                getTypeStr(),
>                                " operation with dynamic rank. Operation name: ",
>                                getName());
>             }
> 
>             bool isScalar = shape.rank().get_length() == 0;
>             outputShapes.emplace_back(isScalar ? ov::PartialShape{1} : shape);
>             originalOutputPrecisions.emplace_back(op->get_output_element_type(i));
>         }
> 
>         childEdges.reserve(outputShapes.size());
>     }
> ```
> 
> Here is the inputs/outputs shapes printed by test_masked_select:
> 
> ```
> self = <Core: available plugins[CPU]>
> model = <Model: 'Model69'
> inputs[
> <ConstOutput: names[x] shape[1,10] type: i32>,
> <ConstOutput: names[mask] shape[1,10] type: boolean>
> ]
> outputs[
> <ConstOutput: names[] shape[...] type: i32>
> ]>
> device_name = 'CPU', config = None
> ```

@mvafin what do you think about it?

File: src/frontends/pytorch/src/utils.cpp [hunk 3]
```diff
@@ -591,6 +592,13 @@ Output<Node> masked_fill(ov::pass::NodeRegistry& rg,
     return rg.make<opset10::Select>(bool_mask, _value, data);
 }
 
+Output<Node> masked_select(ov::pass::NodeRegistry& rg,
+                         const Output<Node>& data,
+                         const Output<Node>& mask) {
+    auto _index = rg.make<opset10::NonZero>(mask);
+    return rg.make<opset10::GatherND>(data, _index);
```
Comment (mvafin) [inline src/frontends/pytorch/src/utils.cpp:599]: You need to do `Transpose` after `NonZero`, since the format of indices in `GatherND` is different. Like here:
https://github.com/openvinotoolkit/openvino/blob/c20d2327316162667ea1f92f947a2bed6303fc4e/src/frontends/pytorch/src/op/index.cpp#L241
